### ISSUE TITLE
Error in calculating Power Output - Too High

### ISSUE THREAD

**Issue by SBFspot** [SBFspot/SBFspot#110]
I've been running SBFspot 3.03 now for a couple weeks, everything has been working fine until the other day.  PVoutput.org stopped receiving my data.  I checked and found that all the data is being recorded in the csv files but just not uploaded to PVoutput.org.  The Log files had the following error:

[22:41:27] ERROR: Uploading 30 datapoints, starting with 20150415,13:40,21454660,5868,,,,124.3,,,,0,2040,0 Bad request 400: Power value [5868] too high for system size [3360]

Checking the csv files, there was a data point that was too high, I have a 3.36kW system and there was a 5.87kW data point at 13:40.  Looking at the data closely, turns out that datapoint was after a 15 minute interval instead of the standard 5 minute interval.  I guess the Raspberry Pi had trouble reaching the inverter for a little while.  The problem is that SBFspot calculated the power output level using the 5 minute interval, causing the output power to be reported 3x higher than it should have been.  Is there a way to have the software calculate power levels based on the actual interval of the data instead of the expected interval?

Thanks,

Pete B

#### Attachments
[Bartplant-20150415.csv](https://www.codeplex.com/Download/AttachmentDownload.ashx?ProjectName=SBFspot&WorkItemId=105&FileAttachmentId=1449727)
[Bartplant-Spot-20150415.csv](https://www.codeplex.com/Download/AttachmentDownload.ashx?ProjectName=SBFspot&WorkItemId=105&FileAttachmentId=1449822)
[Bartplant-User-Events-201504-201504.csv](https://www.codeplex.com/Download/AttachmentDownload.ashx?ProjectName=SBFspot&WorkItemId=105&FileAttachmentId=1449830)
[my-stderr.txt](https://www.codeplex.com/Download/AttachmentDownload.ashx?ProjectName=SBFspot&WorkItemId=105&FileAttachmentId=1449831)
[BartPlant-20150415.csv](https://www.codeplex.com/Download/AttachmentDownload.ashx?ProjectName=SBFspot&WorkItemId=105&FileAttachmentId=1449834)
[my-stdout.txt](https://www.codeplex.com/Download/AttachmentDownload.ashx?ProjectName=SBFspot&WorkItemId=105&FileAttachmentId=1450329)

#### This work item was migrated from CodePlex
CodePlex work item ID: '105'
Vote count: '1'

**Comment by SBFspot** [comment]
[SBF@17/04/2015]
Seems to be related to [this](https://sbfspot.codeplex.com/discussions/628857) one...
Your inverter is missing these entries and I wonder how Sunny Explorer is handling this issue. Can you provide me the SE CSV?

**Comment by SBFspot** [comment]
[SillieWimons@17/04/2015]
dear pete b,    in addition to the question raised by sbf

can you provide the following information:
-a- info about the material that is part of your installation:   eg inverter, other equipment

-b- execute the following command  -  and deliver the output csv files   and the txt file
 !!! while sending the output .csv files to a different directory then the one 
     for your operational system (eg use a different cfg file)
     be aware on the directory for the events 

sbfspot -d5 -v5 -ae5 -ad9 -am2   1>my-stdout.txt 2>my-stderr.txt

rational: 
- there might be several causes for the time jump in your .csv file

eg: 
-d.1- one of them is a shift of the inverter time - this is visible in the events history (cfr option ae5)
-d.2- the option -ad9 will catch the daily history   for the past 9days 
        (this assumes that you will execute this before 23april)
-d.3-  d5 and v5 will allow to get the effictive output by the inverter

kr wim

**Comment by SBFspot** [comment]
[SillieWimons@17/04/2015]
dear pete b,    a question that was not included in my previous note

!!! since you hace csv enabled   you should also have   a   spot table available
??? can you include the spot .csv table, at least for 15april, as part of the info

rational: the spot table might include additional details

kr wim

**Comment by SBFspot** [comment]
[pbartko@17/04/2015]
Okay, attached is the spot file.  I'll work on getting the other files next.

Thanks,

Pete

**Comment by SBFspot** [comment]
[pbartko@17/04/2015]
The my-stdout.txt file is empty, should it have anything in it?

**Comment by SBFspot** [comment]
[pbartko@17/04/2015]
Had to reinstall Sunny Explorer.  Attached is the SE csv file  It shows a few data holes that were not recorded by the Inverter.  Time is shifted by 1 hour, it was not on daylight savings time.

The inverter is a SMA SB3000US model with the SMA bluetooth PiggyBack module installed.  I also have a Sunny Beam monitor that records data from the inverter.

Thanks,

Pete

**Comment by SBFspot** [comment]
[pbartko@17/04/2015]
I took a look at the database.  It looks like the spot values are correct:

sqlite> select * from SpotData where TimeStamp = 1429119607;
1429119607|2000976011|2040|0|7.908|0.0|258.0|0.0|1004|0|0|8.082|0.0|0.0|124.3|125.2|0.0|8449|21454660|60.01|21339.7|20854.2|63.9216|OK|?|0.0

It shows a value of 2040 W for the output power.

However, the vwPvoData View has the following entry:

sqlite> select * from vwPvoData where Timestamp = '2015-04-15 13:40:00';
2015-04-15 13:40:00|SN: 2000976011|SB 3000US|2000976011|21454660|5868||||124.3||||0.0|2040.0|0.0|

It has a value of 5868 for the output power.

**Comment by SBFspot** [comment]
[SillieWimons@18/04/2015]
dear pete,    as you find in SE output, the same records are absent in the tables from SE and SBFspot
 - the difference in the time stamp (SE is one hour different from SBFspot)

!! you have a third source of info via SunnyBeam (these tables are only / 10minutes)
please control whether that info shows the "missing records"

you did not include the stdout file (the stderr file points to a script "bash" fault)
in stdout.txt  you will find:
- whether the inverter has records available for the "holes"   it's a length file but it reads quite simple 
- your config file parameters

another type of info that you can validate with the std-out is whether the clock of
your linux system and the inverter are nearly identical
in your set-up the inverter-clock is "early-morning" set by your SunnyBeam  (within 5minutes after wakeup)
  
note: the first record from your inverter (time 06h35m) is visible via SE - SBFspot hiddes this record

warning: 
- time stamp in the daily file is "inverter-time",  
- time-stamp for spot table ???   this is part of your config file  (system or inverter??)

from the actual info: it looks that you inverter did "temporary" not write records into "5min interval table"
- by consulting the std-out file you can control this 

for extra info that might indicate a cause: you consult the "installer" events
or via SunExplorer (login as installer - if password is not changed it is "1111"
or use SBFspot with the options   -d5 -v5 -ae5 -installer -password1111 (syntac via help)

some warnings/limits that apply on you setup: SunnyBeam  &  SBFspot onLinux  &  SE on ?win? 
-d- from the user-events - you're sunny beam is set for frekwent interaction with the inverter
-d.1- the time updates (event-type 10109 are allways coming from SunBeam susy=99)
-e- inverters (on BT) do only allow two active communications (in parallel)
-e.1- when SE is active the comm's channel stays open for the complete period of usage      
-e.2- SBFspot might fail to communicate if AND sunBeam AND SE is active
this should not be a problem because the next run sbfspot will recuperate as well the previous records

kr wim

**Comment by SBFspot** [comment]
[pbartko@18/04/2015]
I confirmed that the Sunny Beam data files contains the same holes in the data at the same time period.

I re-ran the command:

/usr/local/bin/sbfspot.3/SBFspot -d5 -v5 -ae5 -ad9 -am2 1>my-stdout.txt 2>my-stderr.txt

and received the two output files.  The my-stderr.txt is empty this time.  I've attached the my-stdout.txt file.

I also checked the Events in SE under the installer login.  The only events listed were a time adjust each morning when the inverter wakes up.

Nothing seems to indicate why the inverter failed to record the output data on that day for the missing intervals.  Can anything be done in SBFspot to tolerate missing data like this?

thanks,

Pete

**Comment by SBFspot** [comment]
[pbartko@18/04/2015]
All my data has been rejected from Pvoutput.org since 15 April due to the bad datapoint.  I changed the bad power output data point in the DayData Table of the database and restarted the upload daemon.  At least my data is being uploaded to Pvoutput.org again.  

Pete

**Comment by SBFspot** [comment]
[SBF@18/04/2015]
Since the Sunny Explorer export file proves there is an issue with your inverter, we can only go for the workarounds...
The first one is to simply ignore the power values that are too high. In the db, modify the view vwPvoData by adding a WHERE clause:
``` SQL
SELECT dd.Timestamp,
              dd.Name,
              dd.Type,
              dd.Serial,
              dd.TotalYield AS V1,
              dd.Power AS V2,
              cons.EnergyUsed AS V3,
              cons.PowerUsed AS V4,
              spot.Temperature AS V5,
              spot.Uac1 AS V6,
              NULL AS V7,
              NULL AS V8,
              NULL AS V9,
              NULL AS V10,
              NULL AS V11,
              NULL AS V12,
              dd.PVoutput
         FROM vwDayData AS dd
              LEFT JOIN vwAvgSpotData AS spot
                     ON dd.Serial = spot.Serial AND dd.Timestamp = spot.Nearest5min
              LEFT JOIN vwAvgConsumption AS cons
                     ON dd.Timestamp = cons.Nearest5min
         WHERE dd.Power <= 3360
        ORDER BY dd.Timestamp DESC
```
3360 is the max power of your system. It's hard-coded, and should be changed accordingly.


The second one needs also a modification of the vwPvoData view:
``` SQL
SELECT dd.Timestamp,
              dd.Name,
              dd.Type,
              dd.Serial,
              dd.TotalYield AS V1,
              CASE WHEN dd.Power > 3360 THEN
                  CAST(spot.Pac1 + spot.Pac2 + spot.Pac3 AS INT)
              ELSE
                  dd.Power
              END AS V2,
              cons.EnergyUsed AS V3,
              cons.PowerUsed AS V4,
              spot.Temperature AS V5,
              spot.Uac1 AS V6,
              NULL AS V7,
              NULL AS V8,
              NULL AS V9,
              NULL AS V10,
              NULL AS V11,
              NULL AS V12,
              dd.PVoutput
         FROM vwDayData AS dd
              LEFT JOIN vwAvgSpotData AS spot
                     ON dd.Serial = spot.Serial AND dd.Timestamp = spot.Nearest5min
              LEFT JOIN vwAvgConsumption AS cons
                     ON dd.Timestamp = cons.Nearest5min
        ORDER BY dd.Timestamp DESC
```

In this workaround we take the total Pac power of the spot data when the 5min value is too high. Again the max power is hard-coded.

Tested with SQLite only. SQLite doesn't allow to modify a view, so you have to "DROP" is first and then re-create it.

Give it a try, and let us know please.

**Comment by SBFspot** [comment]
[SillieWimons@20/04/2015]
dear pete,    allow me to add the following comments AFTER reading your std-out

-a- there is another problem, "no data in month archive"
this issue on-top of the problem "missing data in daily file" (skip of records in daily file of 15april)
-a.1- your inverter did not catch data via the "archive-month" command   - see "my-stdout.txt"
BUT there should DEFINITELY be data - at least for "april"
-a.2- this indicates a strange behaviour of your inverter
?? absence of the data can be controlled via SunnyExplorer
!! your sunny beam might hold "older" month tables  ??
info: you can only recuperate data since the installation of the piggy back
!!! upgrade of software in inverter or sunnyBeam might destroy data
 -a.3- please control via the command whether there is something indicated for earlier months
/usr/local/bin/sbfspot.3/SBFspot -d5 -v5 -ae12 -ad0 -am72 -nosql 1>checkmonth-stdout.txt 2>chevkmonth-stderr.txt

-b- since your using the BT-piggy back  (as i do use in my installation)
i suggest that you take the time to control the available data-table at the limits, you best apply two steps:
-b.1- to extract the month tables "if any available"   see my point -a-
/usr/local/bin/sbfspot.3/SBFspot -d5 -v5 -ae12 -ad0 -am72 1>histmon-stdout.txt 2>histmon-stderr.txt
-b.2- to extract the day tables "if any available" 
/usr/local/bin/sbfspot.3/SBFspot -d5 -v5 -ae0 -ad260 -am0 1>histday-stdout.txt 2>histday-stderr.txt

my piggy back in my inverter delivers more then 7 months or  7 * 30 days of history tables
so the above commands might render you meaningfull history info
conditions: clock setting done correctly in the past 
since you're using a sunny-beam that might be ok - which is confirmed by the events list

-c- be aware: 
sunny-beam does not keep indefinite the daily files (in my experience less then 31months)
for that reason i make a back-up of that data each 6months

-d- in your cfg file you apply "SynchTime=1"    that is OK
UNLESS the "delta in time"   between your RPI   and   your SunnyBeam  exceeds   60seconds
in my experience the clock of the Sunny Beam is not very stable
at regular intervals - it is recommended to adjust the clock of the sunny beam
-d.1- from    YOUR  std-out    i'd learn:
Current Inverter Time: 18/04/2015 15:36:27     this time is set by the sunny-beam early morning
your-system-time     -> Sat Apr 18 15:35:49 2015: INFO: Starting...    (cfr std-out)
to-day  time setting by    inverter    and  PCi     differ nearly 40seconds
-d.2- for a "simple" time-compare trick  apply this command   (which 
from the text in the std-out  you can extract:
the system-time-before = "Starting"  the system-time-after     and the current-inverter time
/usr/local/bin/sbfspot.3/SBFspot -d5 -v5 -ae0 -ad0 -am0 -sp0 -nocsv -nosql 1>time-stdout.txt 2>time-stderr.txt

-e- !!!in your std-out I'll found communication warning messages:
-e.1- "Packet ID mismatch. Expected 10, received 185" did appear!!!
this is,   99%certain  :-) , due to the comm's between sunny-beam and inverter
-e.2- Wrong sender: 00:80:25:0B:A3:2C    this is near certain the BT-address of your sunny-beam

**Comment by SBFspot** [comment]
[pbartko@5/05/2015]
I just wanted to provide an update.  I upgraded to V3.07 and have been watching the data for missing values from the inverter.  Today I noticed several holes in the data that would have caused the power output values to be incorrectly reported with V3.05.  The new version, V3.07 correctly reported the power values!

Thanks for the quick update SBF!

Pete

**Comment by SBFspot** [comment]
**Issue closed by _SBF_ with comment**
Fixed in V3.0.7

**Reason closed**
Fixed
